routing_algorithms/map_matching.hpp

```diff
@@ -0,0 +1,627 @@
+                path_lengths.emplace_back(l.size());
+                pruned.emplace_back(l.size());
+            }
+
+            clear(0);
+        }
+
+        void clear(unsigned initial_timestamp)
+        {
+            BOOST_ASSERT(viterbi.size() == parents.size() &&
+                         parents.size() == path_lengths.size() &&
+                         path_lengths.size() == pruned.size() && pruned.size() == breakage.size());
+
+            for (unsigned t = initial_timestamp; t < viterbi.size(); t++)
+            {
+                std::fill(viterbi[t].begin(), viterbi[t].end(), Matching::IMPOSSIBLE_LOG_PROB);
+                std::fill(parents[t].begin(), parents[t].end(), std::make_pair(0u, 0u));
+                std::fill(path_lengths[t].begin(), path_lengths[t].end(), 0);
+                std::fill(pruned[t].begin(), pruned[t].end(), true);
+            }
+            std::fill(breakage.begin() + initial_timestamp, breakage.end(), true);
+        }
+
+        unsigned initialize(unsigned initial_timestamp)
+        {
+            BOOST_ASSERT(initial_timestamp < candidates_list.size());
+
+            do
+            {
+                for (auto s = 0u; s < viterbi[initial_timestamp].size(); ++s)
+                {
+                    viterbi[initial_timestamp][s] =
+                        emission_log_probability(candidates_list[initial_timestamp][s].second);
+                    parents[initial_timestamp][s] = std::make_pair(initial_timestamp, s);
+                    pruned[initial_timestamp][s] =
+                        viterbi[initial_timestamp][s] < Matching::MINIMAL_LOG_PROB;
+
+                    breakage[initial_timestamp] =
+                        breakage[initial_timestamp] && pruned[initial_timestamp][s];
+                }
+
+                ++initial_timestamp;
+            } while (breakage[initial_timestamp - 1]);
+
+            if (initial_timestamp >= viterbi.size())
+            {
+                return Matching::INVALID_STATE;
+            }
+
+            BOOST_ASSERT(initial_timestamp > 0);
+            --initial_timestamp;
+
+            BOOST_ASSERT(breakage[initial_timestamp] == false);
+
+            return initial_timestamp;
+        }
+    };
+
+    // Provides the debug interface for introspection tools
+    struct DebugInfo
+    {
+        DebugInfo(const osrm::json::Logger* logger)
+        : logger(logger)
+        {
+            if (logger)
+            {
+                object = &logger->map->at("matching");
+            }
+        }
+
+        void initialize(const Matching::CandidateLists& candidates_list)
+        {
+            // json logger not enabled
+            if (!logger)
+                return;
+
+            osrm::json::Array states;
+            for (unsigned t = 0; t < candidates_list.size(); t++)
+            {
+                osrm::json::Array timestamps;
+                for (unsigned s = 0; s < candidates_list[t].size(); s++)
```

Comment:
- a range-based for loop with an irange would allow to make `s` const
- `s++` should be `++s`